DEVELOPERS.md

```diff
@@ -0,0 +1,122 @@
+
+This mode is useful in the early stages of lifecycle when incompatible schema changes are made. 
+
+A `DataManipulator` is used to handle the deleting of data. Determining which manipulator to use is handled in the same manner as the [Copier](#copiers), in that there is a `DataManipulatorFactoryManager` which will give a suitable `DataManipulatorFactory` that will return a `DataManipulator` object. 
+
+### Metadata Mirror Replication 
+Only metadata will be copied (mirrored) from the source to the replica. Replica metadata will not be modified so your source and replica will have the same data location.
+
+*NOTE:* The replica table will be marked as `EXTERNAL`. This is done to prevent accidental data loss when dropping the replica. 
+
+For example, this can be used for copying someone else's metadata into your Hive Metastore without copying the data or to replicate a view. You still need to have access to the data in order to query it.
+
+### Metadata Update Replication
+This will update the metadata only for a table that was previously fully replicated.
+
+No data will be copied but any metadata from the source will be copied and table/partition locations will keep pointing to previously replicated data.
+
+Example use case: Update the metadata of a Hive Table (for instance to change the Serde used) without having the overhead of re-replicating all the data.
+
+## Copiers
+The copiers are the classes which do the actual copying of the data. 
+
+There is a `CopierFactoryManager` which determines which type of copier will be used. The `DefaultCopierFactoryManager` is an implementation of this, and has a list of `CopierFactories` auto-wired into it. Spring will find all beans which are implementations of the `CopierFactory` and pass these into the constructor for the `DefaultCopierFactoryManager`. 
+
+There is an optional copier option available to set which `CopierFactory` to use, if this value is set this copier factory class will be used. If this value is not set the `DefaultCopierFactoryManager` will check all `CopierFactories` in the list and return the first which supports replication between the SourceLocation and ReplicaLocation provided. 
+
+There is an order of precedence, which means the `CopierFactories` will be checked in the following order to see if the replication is supported:
+* `S3S3Copier`,
+* `S3MapreduceCpCopier`, 
+* and then falls down to `DistCpCopier` if the above factories don't support the replication.
+
+The copiers which use S3 will create clients that allow access and give permissions to perform actions on S3 buckets. In some cases an IAM role is needed, if data is being transferred across S3 accounts. 
+
+### Types of copier
+**S3S3Copier**
+
+*Replication: S3 → S3* 
+
+This copier uses two `AwsS3Clients` - a source client and a replica client. There is an `AwsS3ClientFactory` which will create clients with the necessary permissions to perform actions on S3 buckets
```

Comment:
Is this true?  When you added `assume-role` to the S3S3Copier, were you able to test assuming a role into the target account?  @KenFigueiredo and I were unable to get it to work - the role being assumed *also* needed S3 READ permissions on the source account in order for us to use S3S3Copier cross-account.  If we were doing something wrong, please let us know, since this is a use-case we are very interested in.

Reply:
I think I did investigate this in the past and you are right. The role does require read permissions on source account. iirc, it was because the S3 class that was copying only uses one client and that was target client. There is no data handover which was possible between two clients.

Reply:
So then this section might need a little tweaking.

Reply:
Hi both, yes it does need read access to the source also. This is mentioned in the readme in the s3-s3 section, but I'll also document it here so its more clear. Thanks for pointing this out!